### PR TITLE
Add docs

### PR DESCRIPTION
**Pull request by adrianchiris**

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -281,6 +293,19 @@ lint-fix: golangci-lint ## Run golangci-lint linter and perform fixes
 generate-mocks: mockery ## generate mock objects
 	PATH=$(LOCALBIN):$(PATH) go generate ./...
 
+
+.PHONY: generate-api-docs
+generate-api-docs: gen-crd-api-reference-docs ## generate api documentation
+	$(GEN_CRD_API_REFERENCE_DOCS) -api-dir=./api/v1alpha1 -config=${CURDIR}/hack/api-docs/config.json \
+	-template-dir=${CURDIR}/hack/api-docs/templates -out-file=$(BUILDDIR)/api-reference.html
+	$(CONTAINER_TOOL) run --rm --volume "`pwd`:/data:Z" pandoc/minimal -f html -t markdown_strict \
```
Comment (adrianchiris) [inline Makefile:301]: if we dont want to rely on docker image of pandoc (hosted at dockerhub) 

an alternative is to:
1. see if we can download binary assuming no dependencies
2. use https://github.com/suntong/html2md

File: docs/api-reference.md [hunk 2]
```diff
@@ -0,0 +1,439 @@
+Packages:
```
Comment (adrianchiris) [inline docs/api-reference.md:1]: html2md generates the followng:



Packages:

- [maintenance.nvidia.com/v1alpha1](#maintenance.nvidia.com%2fv1alpha1)

## maintenance.nvidia.com/v1alpha1

Package v1alpha1 contains API Schema definitions for the maintenance.nvidia.com v1alpha1 API group

Resource Types:

### DrainSpec

( _Appears on:_ [NodeMaintenanceSpec](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceSpec))

DrainSpec describes configuration for node drain during automatic upgrade

FieldDescription`force`

_bool_

Force draining even if there are pods that do not declare a controller

`podSelector`

_string_

PodSelector specifies a label selector to filter pods on the node that need to be drained
For more details on label selectors, see:
[https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)

`timeoutSeconds`

_int32_

TimeoutSecond specifies the length of time in seconds to wait before giving up drain, zero means infinite

`deleteEmptyDir`

_bool_

DeleteEmptyDir indicates if should continue even if there are pods using emptyDir
(local data that will be deleted when the node is drained)

`podEvictionFilters`

_[\[\]PodEvictionFiterEntry](#maintenance.nvidia.com/v1alpha1.PodEvictionFiterEntry)_

PodEvictionFilters specifies filters for pods that need to undergo eviction during drain.
if specified. only pods that match PodEvictionFilters will be evicted during drain operation.
if unspecified. all non-daemonset pods will be evicted.
logical OR is performed between filter entires. logical AND is performed within different filters
in a filter entry.

### DrainStatus

( _Appears on:_ [NodeMaintenanceStatus](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceStatus))

DrainStatus represents the status of draining for the node

FieldDescription`totalPods`

_int32_

TotalPods is the number of pods on the node at the time NodeMaintenance started draining

`evictionPods`

_int32_

EvictionPods is the total number of pods that need to be evicted at the time NodeMaintenance started draining

`drainProgress`

_int32_

DrainProgress represents the draining progress as percentage

`waitForEviction`

_\[\]string_

WaitForEviction is the list of namespaced named pods that need to be evicted

### MaintenanceOperatorConfig

MaintenanceOperatorConfig is the Schema for the maintenanceoperatorconfigs API

FieldDescription`metadata`

_[Kubernetes meta/v1.ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.31/#objectmeta-v1-meta)_
Refer to the Kubernetes API documentation for the fields of the
`metadata` field.
`spec`

_[MaintenanceOperatorConfigSpec](#maintenance.nvidia.com/v1alpha1.MaintenanceOperatorConfigSpec)_

`maxParallelOperations`

_[k8s.io/apimachinery/pkg/util/intstr.IntOrString](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString)_

MaxParallelOperations indicates the maximal number nodes that can undergo maintenance
at a given time. 0 means no limit
value can be an absolute number (ex: 5) or a percentage of total nodes in the cluster (ex: 10%).
absolute number is calculated from percentage by rounding up.
defaults to 1. The actual number of nodes that can undergo maintenance may be lower depending
on the value of MaintenanceOperatorConfigSpec.MaxUnavailable.

`maxUnavailable`

_[k8s.io/apimachinery/pkg/util/intstr.IntOrString](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString)_

MaxUnavailable is the maximum number of nodes that can become unavailable in the cluster.
value can be an absolute number (ex: 5) or a percentage of total nodes in the cluster (ex: 10%).
absolute number is calculated from percentage by rounding up.
by default, unset.
new nodes will not be processed if the number of unavailable node will exceed this value

`logLevel`

_[OperatorLogLevel](#maintenance.nvidia.com/v1alpha1.OperatorLogLevel)_

LogLevel is the operator logging level

`maxNodeMaintenanceTimeSeconds`

_int32_

MaxNodeMaintenanceTimeSeconds is the time from when a NodeMaintenance is marked as ready (phase: Ready)
until the NodeMaintenance is considered stale and removed by the operator.
should be less than idle time for any autoscaler that is running.
default to 30m (1600 seconds)

### MaintenanceOperatorConfigSpec

( _Appears on:_ [MaintenanceOperatorConfig](#maintenance.nvidia.com/v1alpha1.MaintenanceOperatorConfig))

MaintenanceOperatorConfigSpec defines the desired state of MaintenanceOperatorConfig

FieldDescription`maxParallelOperations`

_[k8s.io/apimachinery/pkg/util/intstr.IntOrString](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString)_

MaxParallelOperations indicates the maximal number nodes that can undergo maintenance
at a given time. 0 means no limit
value can be an absolute number (ex: 5) or a percentage of total nodes in the cluster (ex: 10%).
absolute number is calculated from percentage by rounding up.
defaults to 1. The actual number of nodes that can undergo maintenance may be lower depending
on the value of MaintenanceOperatorConfigSpec.MaxUnavailable.

`maxUnavailable`

_[k8s.io/apimachinery/pkg/util/intstr.IntOrString](https://pkg.go.dev/k8s.io/apimachinery/pkg/util/intstr#IntOrString)_

MaxUnavailable is the maximum number of nodes that can become unavailable in the cluster.
value can be an absolute number (ex: 5) or a percentage of total nodes in the cluster (ex: 10%).
absolute number is calculated from percentage by rounding up.
by default, unset.
new nodes will not be processed if the number of unavailable node will exceed this value

`logLevel`

_[OperatorLogLevel](#maintenance.nvidia.com/v1alpha1.OperatorLogLevel)_

LogLevel is the operator logging level

`maxNodeMaintenanceTimeSeconds`

_int32_

MaxNodeMaintenanceTimeSeconds is the time from when a NodeMaintenance is marked as ready (phase: Ready)
until the NodeMaintenance is considered stale and removed by the operator.
should be less than idle time for any autoscaler that is running.
default to 30m (1600 seconds)

### NodeMaintenance

NodeMaintenance is the Schema for the nodemaintenances API

FieldDescription`metadata`

_[Kubernetes meta/v1.ObjectMeta](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.31/#objectmeta-v1-meta)_
Refer to the Kubernetes API documentation for the fields of the
`metadata` field.
`spec`

_[NodeMaintenanceSpec](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceSpec)_

`requestorID`

_string_

RequestorID MUST follow domain name notation format ( [https://tools.ietf.org/html/rfc1035#section-2.3.1](https://tools.ietf.org/html/rfc1035#section-2.3.1))
It MUST be 63 characters or less, beginning and ending with an alphanumeric
character (\[a-z0-9A-Z\]) with dashes (-), dots (.), and alphanumerics between.
caller SHOULD NOT create multiple objects with same requestorID and nodeName.
This field identifies the requestor of the operation.

`additionalRequestors`

_\[\]string_

AdditionalRequestors is a set of additional requestor IDs which are using the same NodeMaintenance
request. addition or removal of requiestor IDs to this list MUST be made with update operation (and retry on failure)
which will replace the entire list.

`nodeName`

_string_

NodeName is The name of the node that maintenance operation will be performed on
creation fails if node obj does not exist (webhook)

`cordon`

_bool_

Cordon if set, marks node as unschedulable during maintenance operation

`waitForPodCompletion`

_[WaitForPodCompletionSpec](#maintenance.nvidia.com/v1alpha1.WaitForPodCompletionSpec)_

WaitForPodCompletion specifies pods via selector to wait for completion before performing drain operation
if not provided, will not wait for pods to complete

`drainSpec`

_[DrainSpec](#maintenance.nvidia.com/v1alpha1.DrainSpec)_

DrainSpec specifies how a node will be drained. if not provided, no draining will be performed.

`status`

_[NodeMaintenanceStatus](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceStatus)_

### NodeMaintenanceSpec

( _Appears on:_ [NodeMaintenance](#maintenance.nvidia.com/v1alpha1.NodeMaintenance))

NodeMaintenanceSpec defines the desired state of NodeMaintenance

FieldDescription`requestorID`

_string_

RequestorID MUST follow domain name notation format ( [https://tools.ietf.org/html/rfc1035#section-2.3.1](https://tools.ietf.org/html/rfc1035#section-2.3.1))
It MUST be 63 characters or less, beginning and ending with an alphanumeric
character (\[a-z0-9A-Z\]) with dashes (-), dots (.), and alphanumerics between.
caller SHOULD NOT create multiple objects with same requestorID and nodeName.
This field identifies the requestor of the operation.

`additionalRequestors`

_\[\]string_

AdditionalRequestors is a set of additional requestor IDs which are using the same NodeMaintenance
request. addition or removal of requiestor IDs to this list MUST be made with update operation (and retry on failure)
which will replace the entire list.

`nodeName`

_string_

NodeName is The name of the node that maintenance operation will be performed on
creation fails if node obj does not exist (webhook)

`cordon`

_bool_

Cordon if set, marks node as unschedulable during maintenance operation

`waitForPodCompletion`

_[WaitForPodCompletionSpec](#maintenance.nvidia.com/v1alpha1.WaitForPodCompletionSpec)_

WaitForPodCompletion specifies pods via selector to wait for completion before performing drain operation
if not provided, will not wait for pods to complete

`drainSpec`

_[DrainSpec](#maintenance.nvidia.com/v1alpha1.DrainSpec)_

DrainSpec specifies how a node will be drained. if not provided, no draining will be performed.

### NodeMaintenanceStatus

( _Appears on:_ [NodeMaintenance](#maintenance.nvidia.com/v1alpha1.NodeMaintenance))

NodeMaintenanceStatus defines the observed state of NodeMaintenance

FieldDescription`conditions`

_[\[\]Kubernetes meta/v1.Condition](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.31/#condition-v1-meta)_

Conditions represents observations of NodeMaintenance current state

`waitForCompletion`

_\[\]string_

WaitForCompletion is the list of namespaced named pods that we wait to complete

`drain`

_[DrainStatus](#maintenance.nvidia.com/v1alpha1.DrainStatus)_

Drain represents the drain status of the node

### OperatorLogLevel ( `string` alias)

( _Appears on:_ [MaintenanceOperatorConfigSpec](#maintenance.nvidia.com/v1alpha1.MaintenanceOperatorConfigSpec))

OperatorLogLevel is the operator log level. one of: \[“debug”, “info”, “error”\]

### PodEvictionFiterEntry

( _Appears on:_ [DrainSpec](#maintenance.nvidia.com/v1alpha1.DrainSpec))

PodEvictionFiterEntry defines filters for Pod evictions during drain operation

FieldDescription`byResourceNameRegex`

_string_

ByResourceNameRegex filters pods by the name of the resources they consume using regex.

### WaitForPodCompletionSpec

( _Appears on:_ [NodeMaintenanceSpec](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceSpec))

WaitForPodCompletionSpec describes the configuration for waiting on pods completion

FieldDescription`podSelector`

_string_

PodSelector specifies a label selector for the pods to wait for completion
For more details on label selectors, see:
[https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#label-selectors)

`timeoutSeconds`

_int32_

TimeoutSecond specifies the length of time in seconds
to wait before giving up on pod termination, zero means infinite

* * *

_Generated with `gen-crd-api-reference-docs`_
_on git commit `5f4aba2`._

File: Makefile (hunk 1 above)
Comment (ykulazhenkov) [reply Makefile:301]: I think docker image is ok

File: README.md [hunk 3]
```diff
@@ -5,7 +5,144 @@
...
+    podSelector: ""
+    timeoutSeconds: 0
+    deleteEmptyDir: true
+    podEvictionFilters:
+    - byResourceNameRegex: nvidia.com/gpu-*
+    - byResourceNameRegex: nvidia.com/rdma*
+
+```
+
+In this example we sequest to perform maintenance for node `worker-1`.
```
Comment (ykulazhenkov) [inline README.md:131]: sequest > request?

File: README.md [hunk 4]
```diff
@@ -5,7 +5,144 @@
...
+
+## Deployment
+
+### Prerequisites
+
+* Kubernetes cluster
+
+### Helm
+
+#### Deploy latest from project sources
```
Comment (ykulazhenkov) [inline README.md:31]: Do you want to add a link to the Helm chart readme to this section?

File: docs/api-reference.md [hunk 5]
```diff
@@ -0,0 +1,439 @@
...
+
+## maintenance.nvidia.com/v1alpha1
+
+Package v1alpha1 contains API Schema definitions for the maintenance.nvidia.com v1alpha1 API group
+
+Resource Types:
+
+### DrainSpec
+
+(*Appears on:*[NodeMaintenanceSpec](#maintenance.nvidia.com/v1alpha1.NodeMaintenanceSpec))
```
Comment (ykulazhenkov) [inline docs/api-reference.md:13]: It looks like these links are broken? Is it possible to fix them somehow?

File: docs/api-reference.md (hunk 2 above)
Comment (ykulazhenkov) [reply docs/api-reference.md:1]: I personally like tables more. I prefer to keep the format that is currently in the PR.

**Comment by ykulazhenkov** [comment]
great work! I added some small comments

File: README.md (hunk 4 above)
Comment (adrianchiris) [reply README.md:31]: i do.

File: README.md (hunk 3 above)
Comment (adrianchiris) [reply README.md:131]: fixed

File: docs/api-reference.md (hunk 5 above)
Comment (adrianchiris) [reply docs/api-reference.md:13]: yea, its not that smooth, lemme try to fix that.

you can specify lua filters to pandoc but im not an experet in neither :\.

**Comment by adrianchiris** [comment]
@ykulazhenkov PTAL,

hacked around with awk to fix references, isnt ideal but IMO it will do for now.

**Pull request merged**